include/nbl/builtin/hlsl/emulated_float64_t.hlsl

```diff
@@ -0,0 +1,596 @@
+                 }
+              }
+           }
+        
+           if (increment)
+           {
+              nbl::hlsl::uint32_t2 added = add64(zFrac0, zFrac1, 0u, 1u);
+              zFrac0 = added.x;
+              zFrac1 = added.y;
+              zFrac1 &= ~((zFrac2 + uint32_t(zFrac2 == 0u)) & uint32_t(roundNearestEven));
+           }
+           else
+           {
+              zExp = LERP(zExp, 0, (zFrac0 | zFrac1) == 0u);
+           }
+           
+           return packFloat64(zSign, zExp, zFrac0, zFrac1);
+        }
+        
+        uint64_t normalizeRoundAndPackFloat64(uint32_t sign, int exp, uint32_t frac0, uint32_t frac1)
+        {
+           int shiftCount;
+           nbl::hlsl::uint32_t3 frac = nbl::hlsl::uint32_t3(frac0, frac1, 0u);
+        
+           if (frac.x == 0u)
+           {
+              exp -= 32;
+              frac.x = frac.y;
+              frac.y = 0u;
+           }
+        
+           shiftCount = countLeadingZeros32(frac.x) - 11;
+           if (0 <= shiftCount)
+           {
+              frac.xy = shortShift64Left(frac.x, frac.y, shiftCount);
+           }
+           else
+           {
+              frac.xyz = shift64ExtraRightJamming(frac.x, frac.y, 0u, -shiftCount);
+           }
+           exp -= shiftCount;
+           return roundAndPackFloat64(sign, exp, frac.x, frac.y, frac.z);
+        }
+
+    }
+
+    struct emulated_float64_t
+    {
+        using storage_t = uint64_t;
+
+        storage_t data;
+
+        // constructors
+        // TODO: specializations
+        template <typename T>
+        static emulated_float64_t create(T val)
+        {
+#ifndef __HLSL_VERSION
+            emulated_float64_t output;
+            output.data = reinterpret_cast<storage_t&>(val);
+            return output;
+#else
+            uint32_t lowBits;
+            uint32_t highBits;
+            asuint(val, lowBits, highBits);
+            
+            emulated_float64_t output;
+            output.data = (uint64_t(highBits) << 32) | uint64_t(lowBits);
+            return output;
+#endif
+        }
+        
+        static emulated_float64_t createEmulatedFloat64PreserveBitPattern(uint64_t val)
+        {
+            emulated_float64_t output;
+            output.data = val;
+            return output;
+        }
+        
+        // TODO: won't not work for uints with msb of index > 52
+#ifndef __HLSL_VERSION
+        template<>
+        static emulated_float64_t create(uint64_t val)
+        {
+#ifndef __HLSL_VERSION
+            const uint64_t msbIndex = nbl::hlsl::findMSB(val);
+#else
+            const uint64_t msbIndex = firstbithigh(val);
+#endif
+            uint64_t exp = ((msbIndex + 1023) << 52) & 0x7FF0000000000000;
+            uint64_t mantissa = (val << (52 - msbIndex)) & 0x000FFFFFFFFFFFFFull;
+            emulated_float64_t output;
+            output.data = exp | mantissa;
+            return output;
+        }
+#endif        

```

Comment:
btw for the `int64_t` "constructor", if the number is negative, negate it and turn into a positive `uint64_t`

construct from uint64_t and if it used to be negative flip the sign bit